File: onebusaway-android/src/main/java/org/onebusaway/android/widealerts/GtfsAlerts.java
package org.onebusaway.android.widealerts;

import com.google.transit.realtime.GtfsRealtime;

import org.onebusaway.android.R;
import org.onebusaway.android.app.Application;
import org.onebusaway.android.util.PreferenceUtils;

import android.content.Context;
import android.util.Log;

import java.net.URL;
import java.util.HashSet;
import java.util.List;
import java.util.Set;

/**
 * Fetches GTFS alerts and processes them.
 */
public class GtfsAlerts {

    private static final String TAG = "GtfsAlerts";
    private static final Set<String> fetchedRegions = new HashSet<>();
    private final Context mContext;

    public GtfsAlerts(Context context) {
        mContext = context;
    }

    /**
     * Fetches GTFS alerts from a specified URL and processes them.
     *
     * @param regionId The current region ID.
     * @param callback The callback to handle the alert data.
     */
    public void fetchAlerts(String regionId, GtfsAlertCallBack callback) {
        if (fetchedRegions.contains(regionId)) {
            Log.d(TAG, "Alerts already fetched for region: " + regionId);
            return;
        }
        String pathUrl = getGtfsAlertsUrl(regionId);
        if (pathUrl == null) {
            return;
        }
        Log.d(TAG, "fetchAlerts for region: " + regionId);
        new Thread(() -> {
            try {
                URL url = new URL(pathUrl);
                GtfsRealtime.FeedMessage feed = GtfsRealtime.FeedMessage.parseFrom(url.openStream());
                // "Now" for the alert start-date window: the feed header timestamp is this feed's
                // server clock (seconds since epoch), so using it cancels device clock skew (#1612).
                // Resolve the device-clock fallback here at the boundary so the downstream check stays
                // a pure function of its inputs.
                long nowMs = feed.hasHeader() && feed.getHeader().hasTimestamp()
                        ? feed.getHeader().getTimestamp() * 1000L
                        : System.currentTimeMillis();
                processAlerts(feed.getEntityList(), nowMs, callback);
                fetchedRegions.add(regionId);
            } catch (Exception e) {
                Log.e(TAG, "Error fetching GTFS alert data for region: " + regionId, e);
                e.printStackTrace();
            }
        }).start();
    }

    /**
     * Processes the list of GTFS alerts and triggers the callback for one valid alert.
     *
     * @param alerts   The list of GTFS alert entities.
     * @param nowMs    "Now" in epoch millis for the start-date window — the feed's server clock, or
     *                 the device clock when the feed carried no header timestamp (resolved by the caller).
     * @param callback The callback to handle each alert.
     */
    public void processAlerts(List<GtfsRealtime.FeedEntity> alerts, long nowMs, GtfsAlertCallBack callback) {
        for (GtfsRealtime.FeedEntity entity : alerts) {
            if (!GtfsAlertsHelper.isValidEntity(mContext, entity, nowMs)) {
                continue;
            }
            GtfsRealtime.Alert alert = entity.getAlert();
            String id = entity.getId();
            String title = GtfsAlertsHelper.getAlertTitle(alert);
            String description = GtfsAlertsHelper.getAlertDescription(alert);
            String url = GtfsAlertsHelper.getAlertUrl(alert);

            Log.d(TAG, "Alert: " + id + " - " + title + " - " + description + " - " + url);
            GtfsAlertsHelper.markAlertAsRead(Application.get().getApplicationContext(), entity);
            callback.onAlert(title, description, url);
            // Only trigger the callback for one alert.
            break;
        }
    }

    /**
     * Constructs the URL for fetching GTFS alerts for a given region.
     *
     * @param regionId The ID of the region for which to fetch alerts.
     * @return The URL to fetch GTFS alerts.
     */
    public String getGtfsAlertsUrl(String regionId) {
        Application app = Application.get();
        String baseUrl = app.getCurrentRegion().getSidecarBaseUrl();
        if (baseUrl == null) return null;
        boolean isTestAlert = PreferenceUtils.getBoolean(app.getString(R.string.preferences_display_test_alerts), false);
        String alertAPIURL = baseUrl + app.getString(R.string.alerts_api_endpoint);
        alertAPIURL = alertAPIURL.replace("regionID", regionId);
        if (isTestAlert) alertAPIURL += "?test=1";
        return alertAPIURL;
    }

}

File: onebusaway-android/src/main/java/org/onebusaway/android/widealerts/GtfsAlerts.kt
package org.onebusaway.android.widealerts

import android.content.Context
import android.util.Log
import com.google.transit.realtime.GtfsRealtime
import dagger.hilt.android.qualifiers.ApplicationContext
import java.net.HttpURLConnection
import java.net.URL
import java.util.concurrent.ConcurrentHashMap
import javax.inject.Inject
import javax.inject.Singleton
import org.onebusaway.android.BuildConfig
import org.onebusaway.android.R
import org.onebusaway.android.app.di.RegionEntryPoint
import org.onebusaway.android.time.ServerTime
import org.onebusaway.android.time.WallTime
import org.onebusaway.android.util.PreferenceUtils

/** Fetches and processes wide GTFS alerts. */
@Singleton
class GtfsAlerts @Inject constructor(
    @param:ApplicationContext private val context: Context
) {
    private val fetchedRegions = ConcurrentHashMap.newKeySet<String>()

    fun fetchAlerts(regionId: String, callback: GtfsAlertCallBack) {
        val pathUrl = getGtfsAlertsUrl(regionId) ?: return
        if (!fetchedRegions.add(regionId)) {
            Log.d(TAG, "Alerts already fetched for region: $regionId")
            return
        }
        if (BuildConfig.DEBUG) Log.d(TAG, "fetchAlerts for region: $regionId")
        Thread {
            try {
                val connection = URL(pathUrl).openConnection() as HttpURLConnection
                connection.connectTimeout = NETWORK_TIMEOUT_MS
                connection.readTimeout = NETWORK_TIMEOUT_MS
                val feed = try {
                    connection.inputStream.use(GtfsRealtime.FeedMessage::parseFrom)
                } finally {
                    connection.disconnect()
                }
                // "Now" for the alert active-window check. The feed header timestamp is this feed's
                // own server clock (GTFS-RT seconds), so preferring it cancels device clock skew
                // (#1612). That field is optional in the spec, and a feed omitting it must still
                // yield alerts, so the device clock is the documented fallback — a deliberate
                // server/device crossing, resolved here at the boundary so the downstream check
                // stays a pure function of its inputs.
                val now = if (feed.hasHeader() && feed.header.hasTimestamp()) {
                    GtfsAlertsHelper.serverTimeFromGtfsSeconds(feed.header.timestamp)
                } else {
                    Log.w(TAG, "GTFS alert feed for region $regionId omitted its timestamp; using the device clock")
                    ServerTime(WallTime.now().epochMs)
                }
                processAlerts(feed.entityList, now, callback)
            } catch (error: Exception) {
                fetchedRegions.remove(regionId)
                Log.e(TAG, "Error fetching GTFS alert data for region: $regionId", error)
            }
        }.start()
    }

    fun processAlerts(
        alerts: List<GtfsRealtime.FeedEntity>,
        now: ServerTime,
        callback: GtfsAlertCallBack
    ) {
        for (entity in alerts) {
            if (!GtfsAlertsHelper.isValidEntity(context, entity, now)) continue
            val alert = entity.alert
            val title = GtfsAlertsHelper.getAlertTitle(alert)
            val description = GtfsAlertsHelper.getAlertDescription(alert)
            val url = GtfsAlertsHelper.getAlertUrl(alert)
            if (BuildConfig.DEBUG) {
                Log.d(TAG, "Alert: ${entity.id} - $title - $description - $url")
            }
            GtfsAlertsHelper.markAlertAsRead(context, entity)
            callback.onAlert(title, description, url)
            break
        }
    }

    fun getGtfsAlertsUrl(regionId: String): String? {
        val region = RegionEntryPoint.get(context).currentRegion() ?: return null
        val baseUrl = region.sidecarBaseUrl ?: return null
        val testAlert = PreferenceUtils.getBoolean(
            context.getString(R.string.preferences_display_test_alerts),
            false
        )
        return (baseUrl + context.getString(R.string.alerts_api_endpoint))
            .replace("regionID", regionId)
            .let { if (testAlert) "$it?test=1" else it }
    }

    private companion object {
        const val TAG = "GtfsAlerts"
        const val NETWORK_TIMEOUT_MS = 15_000
    }
}
